### PR TITLE
add windows support

### PR DESCRIPTION
**Pull request by natefinch**
This replaces the os.Rename call with two separate implementations depending
on target OS.  On windows it uses the movefileex syscall to make the move
atomic, unlike the implementation of os.Rename.

Needs tests, for sure.

### CODE REVIEW

**Comment by natefinch** [comment]
Fixes #1

File: zfile_windows.go [hunk 1]
```diff
@@ -0,0 +1,27 @@
+// mksyscall_windows -l32 file_windows.go
+// MACHINE GENERATED BY THE COMMAND ABOVE; DO NOT EDIT
```
Comment (stapelberg) [inline zfile_windows.go]: Should we have a file with the corresponding go:generate call?

File: file_windows.go [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+//sys moveFileEx(lpExistingFileName *uint16, lpNewFileName *uint16, dwFlags uint32) (err error) = MoveFileExW
+
+// replaceFile atomically replaces the destination file or directory with the
+// source.  It is guaranteed to either replace the target file entirely, or not
+// change either file.
+func replaceFile(source, destination string) error {
+	src, err := syscall.UTF16PtrFromString(source)
+	if err != nil {
+		return &os.LinkError{"replace", source, destination, err}
```
Comment (stapelberg) [inline file_windows.go]: here and below: could you please change the literal construction to use field names to be robust when os.LinkError changes? E.g. `&os.LinkError{Op: "replace", Old: source, New: destination, Err: err}`

File: zfile_windows.go (hunk 1 above)
Comment (natefinch) [reply zfile_windows.go]: That's a possibility... but at the same time, this code should never change, so maybe not?

File: file_windows.go (hunk 2 above)
Comment (natefinch) [reply file_windows.go]: Ahh yeah, definitely.

File: zfile_windows.go (hunk 1 above)
Comment (stapelberg) [reply zfile_windows.go]: Okay. I wasn’t sure under which circumstances this code might change. If the compatibility guarantee is that it keeps working indefinitely, even on new windows versions, I’m fine without the go:generate statement :)

File: file_windows.go (hunk 2 above)
Comment (stapelberg) [reply file_windows.go]: Let me know once you pushed the change :)

File: file_windows.go [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+	}
+	dest, err := syscall.UTF16PtrFromString(destination)
+	if err != nil {
+		return &os.LinkError{"replace", source, destination, err}
+	}
+
+	// see http://msdn.microsoft.com/en-us/library/windows/desktop/aa365240(v=vs.85).aspx
+	if err := moveFileEx(src, dest, movefile_replace_existing|movefile_write_through); err != nil {
```
Comment (tklauser) [inline file_windows.go]: FWIW, `golang.org/x/sys/windows` has a wrapper for `MoveFileExW`, so it could be used instead of generating one here.

**Comment by stapelberg** [comment]
https://github.com/golang/go/issues/32188#issuecomment-497873904 says MoveFileEx is not atomic under windows, and ReplaceFile should instead be used.

**Comment by stapelberg** [comment]
https://github.com/golang/go/issues/22397#issuecomment-498856679 says nothing is atomic under windows. Closing this until someone comes up with new evidence.

**Pull request closed without merging**